File: 1650-lowest-common-ancestor-of-a-binary-tree-iii/1650-lowest-common-ancestor-of-a-binary-tree-iii.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node parent;
};
*/

class Solution {
    public Node lowestCommonAncestor(Node p, Node q) {
        HashSet<Integer> set = new  HashSet<>();
        Node temp1 = p;
        while(temp1!=null) {
            if(temp1==q) return temp1;
            set.add(temp1.val);
            temp1=temp1.parent;
        }
        Node temp2 = q;
        while(temp2!=null) {
            if(temp2==p || set.contains(temp2.val)) return temp2;
            temp2 = temp2.parent;
        }
        return null;
    }
}

File: 1650-lowest-common-ancestor-of-a-binary-tree-iii/1650-lowest-common-ancestor-of-a-binary-tree-iii.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* parent;
};
*/

class Solution {
public:
    Node* lowestCommonAncestor(Node* p, Node * q) {
        Node* temp1 = q;
        unordered_set<int> set;
        //first parent path
        while(temp1) {
            set.insert(temp1->val);
            if(temp1 == p) return temp1;
            temp1 = temp1->parent;
        }
        //second parent path
        temp1 = p;
        while(temp1) {
            if(temp1 == q || set.find(temp1->val)!=set.end()) return temp1;
            temp1 = temp1->parent;
        }
        return nullptr;
    }
};